src/components/cuda/cupti_common.c

```diff
@@ -0,0 +1,465 @@
+    if (papi_errno != PAPI_OK) {
+        return PAPI_EMISC;
+    }
+    else
+        return PAPI_OK;
+}
+
+int util_unload_cuda_sym(void)
+{
+    unload_cuda_sym();
+    unload_cudart_sym();
+    unload_cupti_common_sym();
+    return PAPI_OK;
+}
+
+static int util_dylib_cu_runtime_version(void)
+{
+    int runtimeVersion;
+    CUDART_CALL(cudaRuntimeGetVersionPtr(&runtimeVersion), return PAPI_EMISC );
+    return runtimeVersion;
+}
+
+static int util_dylib_cupti_version(void)
+{
+    unsigned int cuptiVersion;
+    CUPTI_CALL(cuptiGetVersionPtr(&cuptiVersion), return PAPI_EMISC );
+    return cuptiVersion;
+}
+
+int get_device_count(void)
+{
+    static int numDevs = -1;
+    if (numDevs != -1)
+        goto fn_exit;
+    CUDART_CALL(cudaGetDeviceCountPtr(&numDevs), return PAPI_EMISC);
+fn_exit:
+    return numDevs;
+}
+
+static int get_gpu_compute_capability(int dev_num)
+{
+    int cc_major, cc_minor;
+    int cc;
+    CUDART_CALL(cudaDeviceGetAttributePtr(&cc_major,
+        cudaDevAttrComputeCapabilityMajor, dev_num),
+            return PAPI_EMISC );
+    CUDART_CALL(cudaDeviceGetAttributePtr(&cc_minor,
+        cudaDevAttrComputeCapabilityMinor, dev_num),
+            return PAPI_EMISC );
+    cc = cc_major * 10 + cc_minor;
+    return cc;
+}
+
+enum gpu_collection_e {GPU_COLLECTION_UNKNOWN, GPU_COLLECTION_ALL_PERF, GPU_COLLECTION_MIXED, GPU_COLLECTION_ALL_EVENTS, GPU_COLLECTION_ALL_CC70};
+
+static enum gpu_collection_e util_gpu_collection_kind(void)
+{
+    static enum gpu_collection_e kind = GPU_COLLECTION_UNKNOWN;
+    if (kind != GPU_COLLECTION_UNKNOWN)
+        goto fn_exit;
+
+    int total_gpus = get_device_count();
+    int i, cc;
+    int count_perf = 0, count_evt = 0, count_cc70 = 0;
+    for (i=0; i<total_gpus; i++) {
+        cc = get_gpu_compute_capability(i);
+        if (cc == 70) {
+            ++count_cc70;
+        }
+        if (cc >= 70) {
+            ++count_perf;
+        }
+        if (cc <= 70) {
+            ++count_evt;
+        }
+    }
+    if (count_cc70 == total_gpus) {
+        kind = GPU_COLLECTION_ALL_CC70;
+        goto fn_exit;
+    }
+    if (count_perf == total_gpus) {
+        kind = GPU_COLLECTION_ALL_PERF;
+        goto fn_exit;
+    }
+    if (count_evt == total_gpus) {
+        kind = GPU_COLLECTION_ALL_EVENTS;
+        goto fn_exit;
+    }
+    kind = GPU_COLLECTION_MIXED;
+
+fn_exit:
+    return kind;
+}
+
+int cupti_common_init(const char **pdisabled_reason)
+{
+    int papi_errno = util_load_cuda_sym();
+    if (papi_errno != PAPI_OK) {
+        *pdisabled_reason = "Unable to load CUDA library functions.";
+        goto fn_exit;
+    }
+
+    if (util_gpu_collection_kind() == GPU_COLLECTION_MIXED) {
+        *pdisabled_reason = "No support for systems with mixed compute capabilities, such as CC < 7.0 and CC > 7.0 GPUS.";
+        papi_errno = PAPI_ECMP;
+        goto fn_exit;
+    }
+fn_exit:
+    return papi_errno;
+}
+
+int util_runtime_is_perfworks_api(void)
```

Comment:
all functions that are non-static should be namespaced. Use `cuptic_` or `cupti_common_` for those.